lib/spack/spack/paths.py

```diff
@@ -9,135 +9,343 @@
-
-#: installation test (spack test) output
-default_test_path = os.path.join(user_cache_path, "test")
-
-#: spack monitor analysis directories
-default_monitor_path = os.path.join(reports_path, "monitor")
-
-#: git repositories fetched to compare commits to versions
-user_repos_cache_path = os.path.join(user_cache_path, "git_repos")
-
-#: default location where remote package repositories are cloned
-package_repos_path = os.path.join(user_cache_path, "package_repos")
-
-#: bootstrap store for bootstrapping clingo and other tools
-default_user_bootstrap_path = os.path.join(user_cache_path, "bootstrap")
-
-#: transient caches for Spack data (virtual cache, patch sha256 lookup, etc.)
-default_misc_cache_path = os.path.join(user_cache_path, "cache")
-
-#: concretization cache for Spack concretizations
-default_conc_cache_path = os.path.join(default_misc_cache_path, "concretization")
-
-# Below paths pull configuration from the host environment.
-#
-# There are three environment variables you can use to isolate spack from
-# the host environment:
-# - `SPACK_USER_CONFIG_PATH`: override `~/.spack` location (for config and caches)
-# - `SPACK_SYSTEM_CONFIG_PATH`: override `/etc/spack` configuration scope.
-# - `SPACK_DISABLE_LOCAL_CONFIG`: disable both of these locations.
-
-
-# User configuration and caches in $HOME/.spack
-def _get_user_config_path():
-    return os.path.expanduser(os.getenv("SPACK_USER_CONFIG_PATH") or "~%s.spack" % os.sep)
-
-
-# Configuration in /etc/spack on the system
-def _get_system_config_path():
-    return os.path.expanduser(
-        os.getenv("SPACK_SYSTEM_CONFIG_PATH") or os.sep + os.path.join("etc", "spack")
+class XDG_mappings(Enum):
+    config_home = xdg_mapping(
+        spack=XDG_overrides.config_home.value,
+        xdg=XDG_vars.config_home.value,
+        xdg_default=os.path.join("~", ".config"),
+    )
+    state_home = xdg_mapping(
+        spack=XDG_overrides.state_home.value,
+        xdg=XDG_vars.state_home.value,
+        xdg_default=os.path.join("~", ".local", "state"),
+    )
+    data_home = xdg_mapping(
+        spack=XDG_overrides.data_home.value,
+        xdg=XDG_vars.data_home.value,
+        xdg_default=os.path.join("~", ".local", "share"),
+    )
+    cache_home = xdg_mapping(
+        spack=XDG_overrides.cache_home.value,
+        xdg=XDG_vars.cache_home.value,
+        xdg_default=os.path.join("~", ".cache"),
     )
 
 
-#: User configuration location
-user_config_path = _get_user_config_path()
+class Location_vars(Enum):
+    user_cache_path = "USER_CACHE_PATH"
+
+
+# This is for tests that want to clean the environment of XDG_ variables that
+# affect spack behavior
+def _unset_xdg_vars(env):
+    saved = {}
+    for xdg_var in XDG_vars:
+        if xdg_var.value in env:
+            saved[xdg_var.value] = env.pop(xdg_var.value)
+    return saved
+
+
+def _spack_xdg_or_backup(xdg_mapping):
+    if xdg_mapping.spack in os.environ:
+        val = os.environ[xdg_mapping.spack]
```

Comment:
```suggestion
        val = os.path.join(os.environ[xdg_mapping.spack], "spack")
```
Per @becker33's comment, this would be more uniform between xdg and Spack variables. Hmmm now that I see it in code I'm not so sure. I suppose it would make scripting consistent if you switch them, but it does feel redundant from a user perspective too...